java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java

```diff
@@ -0,0 +1,180 @@
+      } else {
+        return ImmutableList.of(programState.addConstraint(operand, new ObjectConstraint(expression, Status.UNLOCKED)));
+      }
+    }
+
+    @Override
+    public String toString() {
+      return super.toString() + ".tryLock()";
+    }
+  }
+
+  private static class TryLockSymbolicValueFactory implements SymbolicValueFactory {
+
+    private final ExpressionTree expression;
+    private final SymbolicValue operand;
+
+    TryLockSymbolicValueFactory(ExpressionTree expression, final SymbolicValue operand) {
+      this.expression = expression;
+      this.operand = operand;
+    }
+
+    @Override
+    public SymbolicValue createSymbolicValue(int counter, Tree syntaxNode) {
+      return new TryLockSymbolicValue(counter, operand, expression);
+    }
+
+  }
+
+  private static class PreStatementVisitor extends CheckerTreeNodeVisitor {
+
+    private static final String LOCK = "java.util.concurrent.locks.Lock";
+    private static final String LOCK_METHOD_NAME = "lock";
+    private static final String TRY_LOCK_METHOD_NAME = "tryLock";
+    private static final String UNLOCK_METHOD_NAME = "unlock";
+
+    private final ConstraintManager constraintManager;
+
+    public PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+      constraintManager = context.getConstraintManager();
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree syntaxNode) {
+      if (syntaxNode.methodSelect().is(Tree.Kind.MEMBER_SELECT)) {
+        MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) syntaxNode.methodSelect();
+        final ExpressionTree expression = memberSelect.expression();
+        if (expression.is(Tree.Kind.IDENTIFIER) && expression.symbolType().isSubtypeOf(LOCK)) {
+          final String methodName = memberSelect.identifier().name();
+          visitMethodInvocationWithIdentifierTarget(methodName, (IdentifierTree) expression);
+        }
+      }
+    }
+
+    private void visitMethodInvocationWithIdentifierTarget(final String methodName, final IdentifierTree target) {
+      if (!isMemberSelectActingOnField(target)) {
+        final SymbolicValue instanceValue = programState.getValue(target.symbol());
+        if (LOCK_METHOD_NAME.equals(methodName)) {
+          track(target, instanceValue, Status.LOCKED);
+        } else if (TRY_LOCK_METHOD_NAME.equals(methodName)) {
+          constraintManager.setValueFactory(new TryLockSymbolicValueFactory(target, instanceValue));
+          track(target, instanceValue, Status.LOCKED);
+        } else if (UNLOCK_METHOD_NAME.equals(methodName)) {
+          track(target, instanceValue, Status.UNLOCKED);
+        }
+      }
+    }
+
+    private static boolean isMemberSelectActingOnField(IdentifierTree expression) {
+      final Symbol symbol = expression.symbol();
+      return ProgramState.isField(symbol);
+    }
+
+    private void track(ExpressionTree syntaxNode, SymbolicValue instanceValue, Status status) {
```

Comment:
why `track` ? why not addConstraint ? and thus ordering parameter with symbolicValue parameter (that should be called like this) first.